### PR TITLE
Support vision datasets for Liger in DPO

### PR DESCRIPTION
**Pull request by albertvillanova**
Support vision datasets for Liger in DPO.

This PR introduces support and testing for vision-language models (VLMs) using the Liger kernel in the DPOTrainer. The main changes include updates to the Liger loss computation to ensure proper handling of vision inputs by invoking the full model backbone instead of just the decoder, and a new test for VLM training with Liger.

Follow-up to review comment: https://github.com/huggingface/trl/pull/5939#discussion_r3356605875

### Motivation

The Liger loss path in `_compute_loss_liger` was calling `model.get_decoder()`, which on VLMs returns only the text decoder and bypasses the vision encoder entirely. Visual inputs (`pixel_values`, `image_grid_thw`, etc.) were also not forwarded.

### Solution

This PR replaces `model.get_decoder()` with `model.model`, the backbone attribute that, for VLMs, includes the vision encoder before the language decoder. This is the same pattern`GRPOTrainer._get_last_hidden_state` already uses, with an explicit comment explaining why `get_decoder()` is wrong for VLMs.

VLM keys are now forwarded using the same exclusion-based `model_kwargs` construction that `_compute_loss` already uses, so any key the processor emits (`pixel_values`, `image_grid_thw`, `pixel_attention_mask`, `image_sizes`, `image_position_ids`, etc.) is forwarded automatically.

### Changes

**Liger Loss Computation Updates:**

* Modified `_compute_loss_liger` in `DPOTrainer` to call the full model backbone (`model.model`) instead of just the decoder, ensuring that vision inputs (e.g., `pixel_values`) are processed correctly for VLMs. This change is applied both for the policy and reference models.
* Cleaned up input handling by filtering out non-model keys and passing all relevant inputs (including vision-related ones) to the model.

**Vision-Language Model (VLM) Support:**

* Added a new test `test_train_vlm_liger` for DPO to verify that VLMs can be trained with the Liger kernel, including checks that model parameters are updated after training.

---

> [!NOTE]
> **Medium Risk**
> Changes core Liger loss forward path for all models (backbone selection and kwargs forwarding); behavior for text-only LMs should be equivalent but VLM and transformers version branches add regression surface.
> 
> **Overview**
> **DPO Liger** now supports vision-language preference training by fixing how hidden states are computed for VLMs.
> 
> `_compute_loss_liger` no longer calls `get_decoder()` with only `input_ids` and `attention_mask`. It builds `model_kwargs` from the batch (excluding DPO-only keys) so vision fields like `pixel_values` and `image_grid_thw` reach the model, and runs the **full multimodal backbone** (`base_model`, or `.model` on transformers &lt; 5.0 for VLMs) for both policy and reference—matching the non-Liger path and GRPO’s VLM pattern.
> 
> A new integration test **`test_train_vlm_liger`** trains a tiny Qwen2.5-VL with `use_liger_kernel=True` and checks loss and parameter updates.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 893a3580bfb8cdec7fca935ee945c97aa9149c4d. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Review by kashif — APPROVED** [review]

File: trl/trainer/dpo_trainer.py [hunk 1]
```diff
@@ -1106,25 +1106,27 @@ def _compute_loss_liger(self, model, inputs, return_outputs):
...
-        completion_mask = inputs["completion_mask"]
+        # Use model.model (backbone including vision encoder) rather than get_decoder() so that VLMs process
+        # pixel_values before the text decoder
+        _non_model_keys = {"completion_mask", "ref_chosen_logps", "ref_rejected_logps"}
+        model_kwargs = {k: v for k, v in inputs.items() if k not in _non_model_keys}
+        model_kwargs["use_cache"] = False
 
-        decoder = model.get_decoder()
-        outputs = decoder(input_ids, attention_mask=attention_mask, use_cache=False)
+        outputs = model.model(**model_kwargs)
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py]: sometimes it's `model.model`, sometimes it's `model.base_model`, I think we should align with https://github.com/huggingface/trl/blob/5f0e83d2bbbad79d6c231164ee09c3492cd27707/trl/trainer/sft_trainer.py#L293-L301

File: trl/trainer/dpo_trainer.py [hunk 2]
```diff
@@ -1106,25 +1106,27 @@ def _compute_loss_liger(self, model, inputs, return_outputs):
...
         hidden_states = outputs.last_hidden_state[:, :-1].contiguous()
         lm_head = model.get_output_embeddings()
         weight = lm_head.weight
         bias = lm_head.bias
 
         with torch.no_grad(), disable_gradient_checkpointing(self.model, self.args.gradient_checkpointing_kwargs):
-            ref_decoder = self.ref_model.get_decoder()
-            ref_outputs = ref_decoder(input_ids, attention_mask=attention_mask, use_cache=False)
+            ref_outputs = self.ref_model.model(**model_kwargs)
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py]: same here

File: tests/test_dpo_trainer.py [hunk 3]
```diff
@@ -1202,3 +1202,31 @@ def test_precompute_ref_log_probs_raises_for_vision(self):
...
+
+    @require_liger_kernel
+    @require_vision
+    def test_train_vlm_liger(self):
+        dataset = load_dataset("trl-internal-testing/zen-image", "conversational_preference", split="train")
+        training_args = DPOConfig(
+            output_dir=self.tmp_dir,
+            max_length=None,
+            per_device_train_batch_size=2,
+            max_steps=3,
```
Comment (qgallouedec) [inline tests/test_dpo_trainer.py]:
```suggestion
            max_length=None,  # for VLMs, truncating can remove image tokens, leading to errors
            per_device_train_batch_size=2,  # VLM training is memory intensive, reduce batch size to avoid OOM
```

for consistency

**Review by qgallouedec — COMMENTED** [review]
a few minor things to address

File: trl/trainer/dpo_trainer.py (hunk 1 above)
Comment (albertvillanova) [reply trl/trainer/dpo_trainer.py]: Super useful insight! Thanks, @qgallouedec! :hugs:

File: tests/test_dpo_trainer.py [hunk 4]
```diff
@@ -1202,3 +1202,31 @@ def test_precompute_ref_log_probs_raises_for_vision(self):
...
+
+    @require_liger_kernel
+    @require_vision
+    def test_train_vlm_liger(self):
+        dataset = load_dataset("trl-internal-testing/zen-image", "conversational_preference", split="train")
+        training_args = DPOConfig(
+            output_dir=self.tmp_dir,
+            max_length=None,  # for VLMs, truncating can remove image tokens, leading to errors
+            per_device_train_batch_size=2,  # VLM training is memory intensive, reduce batch size to avoid OOM
+            max_steps=3,
```
Comment (qgallouedec) [inline tests/test_dpo_trainer.py]: is it needed? we always rely on the dataset len for the test

Comment (albertvillanova) [reply tests/test_dpo_trainer.py]: Thanks, agree: just copied-pasted. I remove it.

**Review by qgallouedec — APPROVED** [review]
